docs/preview/02-Features/06-general-messaging.md

```diff
@@ -9,6 +9,60 @@ Both and future systems also support some general functionality that will be exp
 
 ## Stop message pump when downstream is unable to keep up
 
+### Pause message processing with a circuit breaker
+When your message handler interacts with a dependency on an external resource, that resource may become unavailable. In that case you want to temporarily stop processing messages.
+
+⚠️ This functionality is currently only available for the Azure Service Bus message pump.
+
+⚠️ This functionality is only available when interacting with message pumps, not in message router-only scenarios like Azure Functions.
+
+To interact with the message processing system within your custom message handler, you can inject the `IMessagePumpCircuitBreaker`:
+
+```csharp
+public class OrderMessageHandler : IAzureServiceBusMessageHandler<Order>
+{
+    private readonly IMessagePumpCircuitBreaker _circuitBreaker;
+
+    public OrderMessageHandler(IMessagePumpCircuitBreaker circuitBreaker)
+    {
+        _circuitBreaker = circuitBreaker;
+    }
+
+    public async Task ProcessMessageAsync(Order message, AzureServiceBusMessageContext messageContext,...)
+    {
+        // Determine whether your dependent system is healthy...
+
+        // If not, call the circuit breaker, processing will be halted temporarily.
+        await _circuitBreaker.PauseMessageProcessingAsync(messageContext.JobId);
+    }
+}
+```
+
+The message pump will by default act in the following pattern:
+* Circuit breaker calls `Pause`
+  * Message pump stops processing messages for a period of time (circuit is OPEN).
+* Message pump tries processing a single message (circuit is HALF-OPEN).
+  * Dependency still unhealthy? => Tries again later (circuit is OPEN)
+  * Dependency healthy? => Message pump starts receiving message in full again (circuit is CLOSED).
+
+Both the recovery period after the circuit is open and the interval between messages when the circuit is half-open is configurable when calling the circuit breaker. These time periods are related to your dependent system and could change by the type of transient connection failure.
+
+```csharp
+await _circuitBreaker.PauseMessageProcessingAsync(
+    messageContext.JobId,
+    options =>
+    {
+        // Sets the time period the circuit breaker should wait before retrying to receive messages.
+        // A.k.a. the time period the circuit is closed (default: 10 seconds).
+        options.MessageRecoveryPeriod = TimeSpan.FromSeconds(15);
+
+        // Sets the time period the circuit breaker should wait between each message after the circuit was closed, during recovery.
+        // A.k.a. the time interval to receive messages during which the circuit is half-open (default: 1 second).
+        options.MessageIntervalDuringRecovery = TimeSpan.FromSeconds(1.5);
```

Comment:
Currently only possible to configure the recovery time when the circuit is open and when a single message is being send during the half-open state.
Not yet possible to specify a message count for the amount of message to be sent during this half-open state.

Reply:
I think this is a feature that can be added in a next iteration ?